polytope_server/worker/worker.py

```diff
@@ -211,81 +215,105 @@ def run(self):
-
-        input_data = self.fetch_input_data(request.url)
+        # Creating a new internal span for the full process_request block
+        with create_new_span_internal("Processing request", request.id) as process_span:
+            id = request.id
+            collection = self.collections[request.collection]
 
-        # Dispatch to listed datasources for this collection until we find one that handles the request
-        datasource = None
-        for ds in collection.datasources():
             logging.info(
-                "Processing request using datasource {}".format(ds.get_type()),
+                "Processing request on collection {}".format(collection.name),
                 extra={"request_id": id},
             )
-            if ds.dispatch(request, input_data):
-                datasource = ds
-                break
+            logging.info("Request is: {}".format(request.serialize()))
 
-        # Clean up
-        try:
-            # delete input data if it exists in staging (input data can come from external URLs too)
-            if input_data is not None:
-                if self.staging.query(id):
-                    self.staging.delete(id)
-
-            # upload result data
-            if datasource is not None:
-                request.url = self.staging.create(id, datasource.result(request), datasource.mime_type())
-
-        except Exception as e:
-            request.user_message += f"Failed to finalize request: [{str(type(e))}] {str(e)}"
-            logging.info(request.user_message, extra={"request_id": id})
-            logging.exception("Failed to finalize request", extra={"request_id": id, "exception": str(e)})
-            raise
+            input_data = self.fetch_input_data(request.url)
 
-        # Guarantee destruction of the datasource
-        finally:
-            if datasource is not None:
-                datasource.destroy(request)
+            # Dispatch to listed datasources for this collection until we find one that handles the request
+            datasource = None
+            for ds in collection.datasources():
+                logging.info(
+                    "Processing request using datasource {}".format(ds.get_type()),
+                    extra={"request_id": id},
+                )
 
-        if datasource is None:
-            request.user_message += "Failed to process request."
-            logging.info(request.user_message, extra={"request_id": id})
-            raise Exception("Failed to process request.")
-        else:
-            request.user_message += "Success"
+                # Creating new internal span for the datasource dispatch (calculate roundtrip time)
+                with create_new_span_internal("Datatsource_{}".format(ds.get_type())) as span_ds:
+                    if ds.dispatch(request, input_data):
+                        datasource = ds
+                        span_ds.set_attribute("polytope.datasource", ds.get_type())
```

Comment:
It seems that setting the datasource attribute should be outside of the if block so that we always have this info in the span.

Reply:
This new span is for the dispatch function block, that means that if there is no datasource we will not have a span, therefore, setting the datasource attribute before won't have any effect on the span.

I thought it would be easy to read that setting the span attribute inside of the if it means that the datasource was triggered.

However, If you think it would be better outside then no problem and I move it up

Reply:
This makes sense as is, thanks.